File: src/mVasu.Pwa/src/app/core/models/customer.dto.ts
/**
 * Asiakkaat-tab list row (SCREENS.md §02 + Lumo domain clarification).
 *
 * "Asiakas" maps to one of three XAF entity types under
 * `xVasu.Data.Asma`:
 *
 *   - `Henkilo`        — natural person (Etunimi + Sukunimi)
 *   - `Yritys`         — legal entity (company name + Y-tunnus)
 *   - `Yhteyshenkilo`  — contact person inside a Yritys (firstName +
 *                        lastName + parent Yritys reference)
 *
 * Henkilo and Yritys share an abstract parent (Sopimusasiakas);
 * Sopimus / Hakemus discriminate them at runtime via `OnHenkilo` /
 * `OnYritys` boolean accessors (see Model.xafml AppearanceRules).
 *
 * A customer's "role" is **derived** from related entities, not stored:
 *
 *   - `Hakemus`        — applications / leads, Hakemustyyppi varies
 *   - `SopimusVaraus`  — contract reservations
 *   - `Sopimus`        — active contracts → "asukas / vuokralainen"
 *   - `Tarjous`        — open offers awaiting signature
 *   - `Esittely` /     — scheduled showings
 *     `Tutustumiskaynti`
 *
 * A customer can have any combination — including none at all (rare
 * contact-only records). The counts surfaced on each row let the user
 * triage at a glance without opening the detail.
 *
 * The future BACKEND.md `/api/customers` endpoint should return this
 * exact shape; the mock fixture in CustomersService produces it today.
 */

export type CustomerType = 'person' | 'company' | 'contact-person';

export interface CustomerCounts {
  readonly applications: number;   // Hakemus
  readonly reservations: number;   // SopimusVaraus
  readonly contracts: number;      // Sopimus (active only)
  readonly offers: number;         // Tarjous (status awaiting signature)
  readonly showings: number;       // Esittely / Tutustumiskaynti (upcoming)
}

export type CustomerRelationFilter =
  | 'has-contract'
  | 'has-application'
  | 'has-offer'
  | 'has-showing'
  | 'no-relations';

export interface CustomerTag {
  readonly label: string;
  readonly tone: 'default' | 'cta' | 'info' | 'warn';
}

interface BaseCustomer {
  readonly id: string;
  readonly type: CustomerType;
  /** Used for fi-FI sort and avatar derivation across all types. */
  readonly displayName: string;
  readonly initials: string;
  readonly counts: CustomerCounts;
  readonly primaryAddress?: string;
  readonly city?: string;
  readonly tag?: CustomerTag;
  readonly phone?: string;
  readonly email?: string;

  // Canonical detail-view fields. The list endpoint leaves these undefined;
  // GetById fills them in. Names mirror Asiakas Model.xafml:
  //   PostiNumero, Maa, LangCode, ToimiAla, TyoPaikka, BruttoTulot,
  //   EmailKayttoSallittu, PuhNoKayttoSallittu, Suoramarkkinointikielto,
  //   InfoMessage. PII (PersonID/SSN/DOB/Age) is deliberately not on the
  //   wire.
  readonly postalCode?: string;
  readonly country?: string;
  readonly language?: string;
  readonly industry?: string;
  /** Henkilö-only — Henkilo.Ammatti. */
  readonly profession?: string;
  /** Henkilö-only — Asiakas.TyoPaikka. */
  readonly workplace?: string;
  /** Asiakas.BruttoTulot — bruttotulot €/vuosi. */
  readonly income?: number;
  readonly emailMarketingAllowed?: boolean;
  readonly phoneMarketingAllowed?: boolean;
  readonly directMarketingForbidden?: boolean;
}

export interface PersonCustomer extends BaseCustomer {
  readonly type: 'person';
  readonly firstName: string;       // Henkilo.EtuNimi
  readonly lastName: string;        // Henkilo.Sukunimi
}

export interface CompanyCustomer extends BaseCustomer {
  readonly type: 'company';
  readonly companyName: string;     // Yritys.Nimi
  readonly businessId?: string;     // Y-tunnus
}

export interface ContactPersonCustomer extends BaseCustomer {
  readonly type: 'contact-person';
  readonly firstName: string;
  readonly lastName: string;
  /** Foreign key to the parent Yritys. */
  readonly parentCompanyId: string;
  /** Cached parent company name; rendered as the contact's secondary line. */
  readonly parentCompanyName: string;
}

export type Customer = PersonCustomer | CompanyCustomer | ContactPersonCustomer;

export interface CustomerFilters {
  readonly type?: CustomerType;
  readonly relation?: CustomerRelationFilter;
  readonly city?: string;
}

export const CUSTOMER_TYPE_LABELS: Record<CustomerType, string> = {
  person:           'Henkilöt',
  company:          'Yritykset',
  'contact-person': 'Yhteyshenkilöt',
};

export const CUSTOMER_RELATION_LABELS: Record<CustomerRelationFilter, string> = {
  'has-contract':    'Aktiivinen sopimus',
  'has-application': 'Avoin hakemus',
  'has-offer':       'Avoin tarjous',
  'has-showing':     'Tuleva esittely',
  'no-relations':    'Ei aktiivisia liitoksia',
};

export function totalCount(c: CustomerCounts): number {
  return c.applications + c.reservations + c.contracts + c.offers + c.showings;
}

File: src/mVasu.Api.Contracts/CustomerDto.cs
namespace mVasu.Api.Contracts;

/// <summary>
/// Single customer (Asiakas) row for <c>GET /api/customers</c>. Maps to
/// one of three XAF entity types — <c>Henkilo</c>, <c>Yritys</c> or
/// <c>Yhteyshenkilo</c> (Yritys child). The <see cref="Type"/> field is
/// the discriminator; per-type fields (FirstName/LastName for persons,
/// CompanyName/BusinessId for companies, ParentCompany* for contact
/// persons) are optional at the wire level so the frontend can map
/// straight to its discriminated-union TypeScript shape.
/// </summary>
/// <remarks>
/// <para>"Roles" (asukas / hakija / entinen) are derived, not stored.
/// Each customer carries <see cref="Counts"/> covering active relations
/// — Sopimus / Hakemus / SopimusVaraus / Tarjous / Esittely — and the
/// caller renders them as count pills.</para>
/// </remarks>
public sealed record CustomerDto(
    string Id,
    /// <summary>One of <see cref="CustomerTypes"/>.</summary>
    string Type,
    /// <summary>fi-FI sortable display name. Henkilo / Yhteyshenkilo: "Sukunimi, Etunimi". Yritys: company name.</summary>
    string DisplayName,
    string Initials,
    CustomerCountsDto Counts,
    /// <summary>Active contract address if any (Sopimus.Huoneisto.Osoite).</summary>
    string? PrimaryAddress,
    string? City,
    CustomerTagDto? Tag,
    string? Phone,
    string? Email,
    // Henkilo / Yhteyshenkilo
    string? FirstName,
    string? LastName,
    // Yritys
    string? CompanyName,
    string? BusinessId,
    // Yhteyshenkilo
    string? ParentCompanyId,
    string? ParentCompanyName,
    // -- canonical detail-view fields (per Asiakas Model.xafml) --
    // List view leaves these null; the GetById endpoint populates them
    // when the detail UI calls in.
    /// <summary>Asiakas.PostiNumero — postinumero (Size 5, mask `[0-9]{1,5}`).</summary>
    string? PostalCode = null,
    /// <summary>Asiakas.Maa — maa, oletus piilossa listalla.</summary>
    string? Country = null,
    /// <summary>Asiakas.LangCode — viestintä-kieli (`fi`/`en`/...).</summary>
    string? Language = null,
    /// <summary>Henkilo.Ammatti — vapaa teksti (Size 50). Henkilö-only.</summary>
    string? Profession = null,
    /// <summary>Asiakas.ToimiAla — toimiala (Size 50). Henkilö/Yritys.</summary>
    string? Industry = null,
    /// <summary>Asiakas.TyoPaikka — työnantaja (Size 50). Henkilö-only.</summary>
    string? Workplace = null,
    /// <summary>Asiakas.BruttoTulot — bruttotulot €/vuosi.</summary>
    int? Income = null,
    /// <summary>Asiakas.EmailKayttoSallittu — saako lähettää sähköpostia.</summary>
    bool? EmailMarketingAllowed = null,
    /// <summary>Asiakas.PuhNoKayttoSallittu — saako soittaa.</summary>
    bool? PhoneMarketingAllowed = null,
    /// <summary>Asiakas.Suoramarkkinointikielto — kielto suoramarkkinoinnille.</summary>
    bool? DirectMarketingForbidden = null);
    // NOTE: Asiakas.InfoMessage is XPO-non-persistent and can't be
    // projected through SelectData. Reintroduce it as a derived banner
    // (e.g. "Asiakkaalla on perinnänestoja") computed from the related
    // flags if the UI needs it.

/// <summary>Aggregated active-relation counts. All values are non-negative.</summary>
public sealed record CustomerCountsDto(
    int Applications,   // Hakemus
    int Reservations,   // SopimusVaraus
    int Contracts,      // Sopimus (active only)
    int Offers,         // Tarjous (status awaiting signature)
    int Showings);      // Esittely / Tutustumiskaynti (upcoming)

/// <summary>
/// Optional status badge derived from the customer's relation state —
/// e.g. "Allekirjoitus" when an offer awaits signing, "Päättyy 30.6."
/// when a contract has a termination date within 30 days.
/// </summary>
public sealed record CustomerTagDto(
    string Label,
    /// <summary>One of "default" / "cta" / "info" / "warn".</summary>
    string Tone);

/// <summary>Top-level <c>GET /api/customers</c> envelope.</summary>
public sealed record CustomersResponseDto(
    IReadOnlyList<CustomerDto> Items,
    int Total);

/// <summary>Canonical customer-type discriminator values.</summary>
public static class CustomerTypes
{
    public const string Person        = "person";
    public const string Company       = "company";
    public const string ContactPerson = "contact-person";

    public static readonly IReadOnlySet<string> Valid =
        new HashSet<string> { Person, Company, ContactPerson };
}

/// <summary>Relation filters supported by the <c>relation</c> query param.</summary>
public static class CustomerRelationFilters
{
    public const string HasContract    = "has-contract";
    public const string HasApplication = "has-application";
    public const string HasOffer       = "has-offer";
    public const string HasShowing     = "has-showing";
    public const string NoRelations    = "no-relations";

    public static readonly IReadOnlySet<string> Valid =
        new HashSet<string> { HasContract, HasApplication, HasOffer, HasShowing, NoRelations };
}
